adding functionality to switch to isoAdvector

cosicp opened the pull request.

cosicp committed:
adding optionality to switch to isoAdvector

cosicp committed:
small fixes

cosicp committed:
Deleting settings.json

tomflint22 reviewed (approved):
I love how you have done this!!

philipcardiff committed:
laserbeamFoam: update code style conventions and avoid file name prefixes

philipcardiff committed:
Move laserMeltFoam libraries to src and rename correctPhi.H to
 correctPhi.lowercase.H to allow the solver to compile on case-insensitive
 drives

philipcardiff committed:
Remove work-in-progress compressibleMulticomponentMultistatelaserbeamFoam solver

philipcardiff committed:
Fix for applications/Allwmake

philipcardiff commented:
@cosicp @tomflint22

I made some changes:
- To avoid renaming files, I changed the code to directly refer to the appropriate isoAdvector or MULES, e.g. `#include "isoAdvector/createPorosity.H" rather than `#include "ISOADVECTOR_createPorosity.H"`.
- I removed any files or variables or directories with an "_" and the OpenFOAM convention is to use `camelCase` not `snake_case`
- Moved the laserMeltFoam libraries to the src directory
- Removed the work-in-progress compressibleMulticomponentMultistatelaserbeamFoam solver - this can be added back when it compiles (and includes the relevant libraries)

Comment:
- Currently, much of the isoAdvector initialisations are still performed even if MULES is selected, and vice versa. This is ok for now, but it is not ideal. A proper (OpenFOAM-style) fix for this would be to define a run-time selectable interface class to hold the MULES and isoAdvector code. This is not too much effort and I am happy to do it if people like - it can be a future pull request as the current code is ok for now.

The merge is ok for me now (assuming it passes the build check below). Double-check everything compiles for you and if you are happy we can merge it.

tomflint22 commented:
Thanks Philip! Yeah sounds like a run-time selectable interface class would be the way to go. Not needed right now but something for the not too-distant future I think

I'm just on the train. @cosicp  or @ScimonCFD , can you check this all compiles OK for you. Once this is merged I'll re-add the compressible solver and push that tonight

philipcardiff commented:
The `Alltest` is failing. probably because I renamed the `interface_tracking_scheme` to `interfaceTrackingScheme`. Let me check

~~EDIT: This only explains the laserMeltFoam error. Most of the errors are from the laserbeamFoam cases.~~
EDIT 2: sorry, I confused myself; yep, `interfaceTrackingScheme` needs to be added to the laserBeamFoam cases.

philipcardiff committed:
Fix for laserMeltFoam/LPBF_tutorial: add interfaceTrackingScheme flag

philipcardiff committed:
Fix for new interfaceTrackingScheme flag and remove log files

philipcardiff committed:
Update gitignore

cosicp commented:
> The `Alltest` is failing. probably because I renamed the `interface_tracking_scheme` to `interfaceTrackingScheme`. Let me check
> 
> EDIT: This only explains the laserMeltFoam error. Most of the errors are from the laserbeamFoam cases.

Are the Alltest scripts updated for v2506 or are they the same ones that we had for of10? Do OF versions matter in these tests?

philipcardiff committed:
Fix for new interfaceTrackingScheme flag: move from LaserProperties to transportProperties

philipcardiff commented:
> > The `Alltest` is failing. probably because I renamed the `interface_tracking_scheme` to `interfaceTrackingScheme`. Let me check
> > EDIT: This only explains the laserMeltFoam error. Most of the errors are from the laserbeamFoam cases.
> 
> Are the Alltest scripts updated for v2506 or are they the same ones that we had for of10? Do OF versions matter in these tests?

The Alltest script is basically taken from solids4foam so it should work with ORG and COM; it seems to work as expected, i.e. set cases to run for one time-step/iteration and check they exit cleanly.

I think the laserbeamFoam cases should work now (I was putting the interfaceTrackingScheme in the wrong file).

But laserMeltFoam fails with:

```
Courant Number mean: 0 max: 0
deltaT = 1.2e-12
Time = 1.2e-12

PIMPLE: iteration 1
Setting up velocity and flux fields for isoAdvector.
Selecting reconstructionScheme: isoAlpha
isoAdvection: After conservative bounding: min(alpha) = 0, max(alpha) = 1 + 0
Restoring velocity and flux fields after isoAdvector.
Material phase volume fraction = 0.604381  Min(alphaM) = 0  Max(alphaM) = 1
Laser power: 150
Laser position: (0.0001 0.0001 0)


--> FOAM FATAL ERROR: (openfoam-2506)
Argument of trancendental function not dimensionless


    From Foam::dimensionSet Foam::trans(const dimensionSet &)
    in file dimensionSet/dimensionSet.C at line 482.
```

cosicp commented:
> > > The `Alltest` is failing. probably because I renamed the `interface_tracking_scheme` to `interfaceTrackingScheme`. Let me check
> > > EDIT: This only explains the laserMeltFoam error. Most of the errors are from the laserbeamFoam cases.
> > 
> > 
> > Are the Alltest scripts updated for v2506 or are they the same ones that we had for of10? Do OF versions matter in these tests?
> 
> The Alltest script is basically taken from solids4foam so it should work with ORG and COM; it seems to work as expected, i.e. set cases to run for one time-step/iteration and check they exit cleanly.
> 
> I think the laserbeamFoam cases should work now (I was putting the interfaceTrackingScheme in the wrong file).
> 
> But laserMeltFoam fails with:
> 
> ```
> Courant Number mean: 0 max: 0
> deltaT = 1.2e-12
> Time = 1.2e-12
> 
> PIMPLE: iteration 1
> Setting up velocity and flux fields for isoAdvector.
> Selecting reconstructionScheme: isoAlpha
> isoAdvection: After conservative bounding: min(alpha) = 0, max(alpha) = 1 + 0
> Restoring velocity and flux fields after isoAdvector.
> Material phase volume fraction = 0.604381  Min(alphaM) = 0  Max(alphaM) = 1
> Laser power: 150
> Laser position: (0.0001 0.0001 0)
> 
> 
> --> FOAM FATAL ERROR: (openfoam-2506)
> Argument of trancendental function not dimensionless
> 
> 
>     From Foam::dimensionSet Foam::trans(const dimensionSet &)
>     in file dimensionSet/dimensionSet.C at line 482.
> ```

I think it might have something to do with the units of things in exp()in the old laser model that laserBeamFoam has.

philipcardiff commented:
> I think it might have something to do with the units of things in exp()in the old laser model that laserBeamFoam has.

OK, I will let you figure this one out; let me know if you need help. The error basically says that you are trying to take an exp/sin/cos/etc of a non-dimensionless number, which doesn't make sense.

philipcardiff commented:
At least, all the laserbeamFoam cases now pass; so once you fix this laserMeltFoam case, then the test will pass.

tomflint22 commented:
@cosicp , any updates on fixing the units issue? Once this is merged I'll open up a new PR for the compressible solver...... before I lose all the files again!

cosicp commented:
Yes, I have figured out the problem a little while ago but ive been trying to update the other things on laserMeltFoam before I push it but I guess its better to fix this and make a new pull request for the rest after. 

I am not sure how to follow-up on Philips commits so I can write out the line that needs to be changed here so @philipcardiff can just update and merge it.

The problem was in the tutorial case actually. When the radius of the laser is read from the LaserProperties the old model assumes it is a dimensioned scalar and the new laser model with the ray-tracing just reads its value.

So if you specify the radius in constant/LaserProperties as  **laserRadius  laserRadius  [0 1 0 0 0 0] 25e-6;** the old model will be happy but the ray tracing will break and if you specify it as **laserRadius 25e-6;**  the ray tracing will be fine but the old model will break (which is what is happening now). 

This is a simple fix, @philipcardiff you can just add the dimensions in the line 41 of readLaserProperties.H where the radius is read from the input file:

from:
**const dimensionedScalar r0("laserRadius", LaserProperties); // Laser radius**
to:
**const dimensionedScalar r0("laserRadius",dimensionSet(0, 1, 0, 0, 0, 0, 0), LaserProperties); // Laser radius**

and then 

laserRadius 25e-6;

will be fine for both laser models.

philipcardiff commented:
OK, thanks, I will make the change

philipcardiff committed:
Update laserMeltFoam/readLaserProperties.H per https://github.com/laserbeamfoam/LaserbeamFoam/pull/78

philipcardiff commented:
All tests passed, thanks, Petar. This is now merged!
